<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Getting started with the plyranges package â€¢ plyranges</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/journal/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Getting started with the plyranges package">
<meta property="og:description" content="plyranges">
<meta property="og:image" content="/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">plyranges</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.11.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/an-introduction.html">Introduction to plyranges</a>
    </li>
  </ul>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/sa-lee/plyranges">
    <span class="fas fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="an-introduction_files/header-attrs-2.7/header-attrs.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Getting started with the plyranges package</h1>
                        <h4 class="author">Stuart Lee</h4>
            
            <h4 class="date">2021-05-06</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/sa-lee/plyranges/blob/master/vignettes/an-introduction.Rmd"><code>vignettes/an-introduction.Rmd</code></a></small>
      <div class="hidden name"><code>an-introduction.Rmd</code></div>

    </div>

    
    
<div id="ranges-revisited" class="section level1">
<h1 class="hasAnchor">
<a href="#ranges-revisited" class="anchor"></a><code>Ranges</code> revisited</h1>
<p>In Bioconductor there are two classes, <code>IRanges</code> and <code>GRanges</code>, that are standard data structures for representing genomics data. Throughout this document I refer to either of these classes as <code>Ranges</code> if an operation can be performed on either class, otherwise I explicitly mention if a function is appropriate for an <code>IRanges</code> or <code>GRanges</code>.</p>
<p><code>Ranges</code> objects can either represent sets of integers as <code>IRanges</code> (which have start, end and width attributes) or represent genomic intervals (which have additional attributes, sequence name, and strand) as <code>GRanges</code>. In addition, both types of <code>Ranges</code> can store information about their intervals as metadata columns (for example GC content over a genomic interval).</p>
<p><code>Ranges</code> objects follow the tidy data principle: each row of a <code>Ranges</code> object corresponds to an interval, while each column will represent a variable about that interval, and generally each object will represent a single unit of observation (like gene annotations).</p>
<p>Consequently, <code>Ranges</code> objects provide a powerful representation for reasoning about genomic data. In this vignette, you will learn more about <code>Ranges</code> objects and how via grouping, restriction and summarisation you can perform common data tasks.</p>
</div>
<div id="constructing-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#constructing-ranges" class="anchor"></a>Constructing <code>Ranges</code>
</h1>
<p>To construct an <code>IRanges</code> we require that there are at least two columns that represent at either a starting coordinate, finishing coordinate or the width of the interval.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/base/message.html">suppressPackageStartupMessages</a></span><span class="op">(</span><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">plyranges</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/Random.html">set.seed</a></span><span class="op">(</span><span class="fl">100</span><span class="op">)</span>
<span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">:</span><span class="op">-</span><span class="fl">1</span>, <span class="fl">13</span><span class="op">:</span><span class="fl">15</span><span class="op">)</span>,
                 width<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">0</span><span class="op">:</span><span class="fl">3</span>, <span class="fl">2</span><span class="op">:</span><span class="fl">0</span><span class="op">)</span><span class="op">)</span>

<span class="co"># produces IRanges</span>
<span class="va">rng</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/as_iranges.html">as_iranges</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">rng</span></code></pre></div>
<pre><code>## IRanges object with 7 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         2         1         0
##   [2]         1         1         1
##   [3]         0         1         2
##   [4]        -1         1         3
##   [5]        13        14         2
##   [6]        14        14         1
##   [7]        15        14         0</code></pre>
<p>To construct a <code>GRanges</code> we require a column that represents that sequence name ( contig or chromosome id), and an optional column to represent the strandedness of an interval.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># seqname is required for GRanges, metadata is automatically kept</span>
<span class="va">grng</span> <span class="op">&lt;-</span> <span class="va">df</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://rdrr.io/r/base/transform.html">transform</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>, <span class="fl">7</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
         strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span>, <span class="fl">7</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
         gc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">7</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/as_iranges.html">as_granges</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">grng</span></code></pre></div>
<pre><code>## GRanges object with 7 ranges and 1 metadata column:
##       seqnames    ranges strand |        gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##   [1]     chr2       2-1      - |  0.762551
##   [2]     chr1         1      - |  0.669022
##   [3]     chr2       0-1      + |  0.204612
##   [4]     chr2      -1-1      - |  0.357525
##   [5]     chr1     13-14      - |  0.359475
##   [6]     chr1        14      - |  0.690291
##   [7]     chr2     15-14      - |  0.535811
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="arithmetic-on-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#arithmetic-on-ranges" class="anchor"></a>Arithmetic on Ranges</h1>
<p>Sometimes you want to modify a genomic interval by altering the width of the interval while leaving the start, end or midpoint of the coordinates unaltered. This is achieved with the <code>mutate</code> verb along with <code>anchor_*</code> adverbs.</p>
<p>The act of anchoring fixes either the start, end, center coordinates of the <code>Range</code> object, as shown in the figure and code below and anchors are used in combination with either <code>mutate</code> or <code>stretch</code>.</p>
<p><img src="anchors.png" width="400px" style="display: block; margin: auto;"></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rng</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_iranges.html">as_iranges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, end<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="va">grng</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/as_iranges.html">as_granges</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span>, <span class="fl">3</span><span class="op">)</span>, end<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">2</span>, <span class="fl">8</span><span class="op">)</span>,
                          seqnames <span class="op">=</span> <span class="st">"seq1"</span>,
                          strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"*"</span>, <span class="st">"-"</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>
<span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="va">rng</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1        10        10
##   [2]         2        11        10
##   [3]         3        12        10</code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_start</a></span><span class="op">(</span><span class="va">rng</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         1        10        10
##   [2]         2        11        10
##   [3]         3        12        10</code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_end</a></span><span class="op">(</span><span class="va">rng</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -4         5        10
##   [2]        -7         2        10
##   [3]        -1         8        10</code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_center</a></span><span class="op">(</span><span class="va">rng</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -2         7        10
##   [2]        -3         6        10
##   [3]         1        10        10</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_3p</a></span><span class="op">(</span><span class="va">grng</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co"># leave negative strand fixed</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1      -4-5      +
##   [2]     seq1      -7-2      *
##   [3]     seq1      3-12      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_5p</a></span><span class="op">(</span><span class="va">grng</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="co"># leave positive strand fixed</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1      1-10      +
##   [2]     seq1      2-11      *
##   [3]     seq1      -1-8      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Similarly, you can modify the width of an interval using the <code>stretch</code> verb. Without anchoring, this function will extend the interval in either direction by an integer amount. With anchoring, either the start, end or midpoint are preserved.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">rng2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/stretch.html">stretch</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_center</a></span><span class="op">(</span><span class="va">rng</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span>
<span class="va">rng2</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -4        10        15
##   [2]        -3         7        11
##   [3]        -2        13        16</code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/stretch.html">stretch</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_end</a></span><span class="op">(</span><span class="va">rng2</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]       -14        10        25
##   [2]       -13         7        21
##   [3]       -12        13        26</code></pre>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/stretch.html">stretch</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_start</a></span><span class="op">(</span><span class="va">rng2</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -4        20        25
##   [2]        -3        17        21
##   [3]        -2        23        26</code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/stretch.html">stretch</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_3p</a></span><span class="op">(</span><span class="va">grng</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1      -9-5      +
##   [2]     seq1      -8-2      *
##   [3]     seq1      3-18      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/stretch.html">stretch</a></span><span class="op">(</span><span class="fu"><a href="../reference/anchor.html">anchor_5p</a></span><span class="op">(</span><span class="va">grng</span><span class="op">)</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1      1-15      +
##   [2]     seq1      2-12      *
##   [3]     seq1      -7-8      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p><code>Ranges</code> can be shifted left or right. If strand information is available we can also shift upstream or downstream.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/shift_left.html">shift_left</a></span><span class="op">(</span><span class="va">rng</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        -9        -5         5
##   [2]        -8        -8         1
##   [3]        -7        -2         6</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/shift_left.html">shift_right</a></span><span class="op">(</span><span class="va">rng</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 3 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        11        15         5
##   [2]        12        12         1
##   [3]        13        18         6</code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/shift_left.html">shift_upstream</a></span><span class="op">(</span><span class="va">grng</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1     -9--5      +
##   [2]     seq1        -8      *
##   [3]     seq1     13-18      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/shift_left.html">shift_downstream</a></span><span class="op">(</span><span class="va">grng</span>, <span class="fl">10</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]     seq1     11-15      +
##   [2]     seq1        12      *
##   [3]     seq1     -7--2      -
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="grouping-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#grouping-ranges" class="anchor"></a>Grouping <code>Ranges</code>
</h1>
<p><code>plyranges</code> introduces a new class of <code>Ranges</code> called <code>RangesGrouped</code>, this is a similar idea to the grouped <code>data.frame\tibble</code> in <code>dplyr</code>.</p>
<p>Grouping can act on either the core components or the metadata columns of a <code>Ranges</code> object.</p>
<p>It is most effective when combined with other verbs such as <code><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code>, <code><a href="https://dplyr.tidyverse.org/reference/filter.html">filter()</a></code>, <code><a href="../reference/reduce_ranges.html">reduce_ranges()</a></code> or <code><a href="../reference/disjoin_ranges.html">disjoin_ranges()</a></code>.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grng</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"chr1"</span>, <span class="st">"chr2"</span><span class="op">)</span>, <span class="fl">7</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
         strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span>, <span class="fl">7</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span>,
         gc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fl">7</span><span class="op">)</span>,
         start <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">7</span>,
         width <span class="op">=</span> <span class="fl">10</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/as_iranges.html">as_granges</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">grng_by_strand</span> <span class="op">&lt;-</span> <span class="va">grng</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">strand</span><span class="op">)</span>

<span class="va">grng_by_strand</span></code></pre></div>
<pre><code>## GRanges object with 7 ranges and 1 metadata column:
## Groups: strand [2]
##       seqnames    ranges strand |        gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##   [1]     chr2      1-10      - |  0.889454
##   [2]     chr2      2-11      + |  0.180407
##   [3]     chr1      3-12      - |  0.629391
##   [4]     chr2      4-13      + |  0.989564
##   [5]     chr1      5-14      - |  0.130289
##   [6]     chr1      6-15      - |  0.330661
##   [7]     chr2      7-16      - |  0.865121
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
</div>
<div id="restricting-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#restricting-ranges" class="anchor"></a>Restricting <code>Ranges</code>
</h1>
<p>The verb <code>filter</code> can be used to restrict rows in the <code>Ranges</code>. Note that grouping will cause the <code>filter</code> to act within each group of the data.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grng</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gc</span> <span class="op">&lt;</span> <span class="fl">0.3</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |        gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##   [1]     chr2      2-11      + |  0.180407
##   [2]     chr1      5-14      - |  0.130289
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="co"># filtering by group</span>
<span class="va">grng_by_strand</span> <span class="op">%&gt;%</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span><span class="op">(</span><span class="va">gc</span> <span class="op">==</span> <span class="fu"><a href="https://rdrr.io/r/base/Extremes.html">max</a></span><span class="op">(</span><span class="va">gc</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
## Groups: strand [2]
##       seqnames    ranges strand |        gc
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;numeric&gt;
##   [1]     chr2      1-10      - |  0.889454
##   [2]     chr2      4-13      + |  0.989564
##   -------
##   seqinfo: 2 sequences from an unspecified genome; no seqlengths</code></pre>
<p>We also provide the convenience methods <code>filter_by_overlaps</code> and <code>filter_by_non_overlaps</code> for restricting by any overlapping <code>Ranges</code>.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ir0</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">5</span>,<span class="fl">10</span>, <span class="fl">15</span>,<span class="fl">20</span><span class="op">)</span>, width <span class="op">=</span> <span class="fl">5</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/as_iranges.html">as_iranges</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">ir1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>start <span class="op">=</span> <span class="fl">2</span><span class="op">:</span><span class="fl">6</span>, width <span class="op">=</span> <span class="fl">3</span><span class="op">:</span><span class="fl">7</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/as_iranges.html">as_iranges</a></span><span class="op">(</span><span class="op">)</span>
<span class="va">ir0</span></code></pre></div>
<pre><code>## IRanges object with 4 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5
##   [2]        10        14         5
##   [3]        15        19         5
##   [4]        20        24         5</code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ir1</span></code></pre></div>
<pre><code>## IRanges object with 5 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         2         4         3
##   [2]         3         6         4
##   [3]         4         8         5
##   [4]         5        10         6
##   [5]         6        12         7</code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ir0</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/filter_by_overlaps.html">filter_by_overlaps</a></span><span class="op">(</span><span class="va">ir1</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]         5         9         5
##   [2]        10        14         5</code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ir0</span> <span class="op">%&gt;%</span> <span class="fu"><a href="../reference/filter_by_overlaps.html">filter_by_non_overlaps</a></span><span class="op">(</span><span class="va">ir1</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 2 ranges and 0 metadata columns:
##           start       end     width
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt;
##   [1]        15        19         5
##   [2]        20        24         5</code></pre>
</div>
<div id="summarising-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#summarising-ranges" class="anchor"></a>Summarising <code>Ranges</code>
</h1>
<p>The <code>summarise</code> function will return a <code>DataFrame</code> because the information required to return a <code>Ranges</code> object is lost. It is often most useful to use <code><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise()</a></code> in combination with the <code><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by()</a></code> family of functions.</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ir1</span> <span class="op">&lt;-</span> <span class="va">ir1</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>gc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/length.html">length</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>

<span class="va">ir0</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/find_overlaps.html">group_by_overlaps</a></span><span class="op">(</span><span class="va">ir1</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>gc <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">gc</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 2 rows and 2 columns
##       query        gc
##   &lt;integer&gt; &lt;numeric&gt;
## 1         1  0.675555
## 2         2  0.635795</code></pre>
</div>
<div id="joins-or-another-way-at-looking-at-overlaps-between-ranges" class="section level1">
<h1 class="hasAnchor">
<a href="#joins-or-another-way-at-looking-at-overlaps-between-ranges" class="anchor"></a>Joins, or another way at looking at overlaps between <code>Ranges</code>
</h1>
<p>A join acts on two GRanges objects, a query and a subject.</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">query</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"chr1"</span>,
               strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"+"</span>, <span class="st">"-"</span><span class="op">)</span>,
               start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">9</span><span class="op">)</span>,
               end <span class="op">=</span>  <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">7</span>, <span class="fl">10</span><span class="op">)</span>,
               key.a <span class="op">=</span> <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/as_iranges.html">as_granges</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">subject</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"chr1"</span>,
               strand <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"-"</span>, <span class="st">"+"</span><span class="op">)</span>,
               start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">6</span><span class="op">)</span>,
               end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">8</span><span class="op">)</span>,
               key.b <span class="op">=</span> <span class="va">LETTERS</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/as_iranges.html">as_granges</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<div class="figure">
<img src="an-introduction_files/figure-html/olap-1.png" alt="Query and Subject Ranges" width="700"><p class="caption">
Query and Subject Ranges
</p>
</div>
<p>The join operator is relational in the sense that metadata from the query and subject ranges is retained in the joined range. All join operators in the <code>plyranges</code> DSL generate a set of hits based on overlap or proximity of ranges and use those hits to merge the two datasets in different ways. There are four supported matching algorithms: <em>overlap</em>, <em>nearest</em>, <em>precede</em>, and <em>follow</em>. We can further restrict the matching by whether the query is completely <em>within</em> the subject, and adding the <em>directed</em> suffix ensures that matching ranges have the same direction (strand).</p>
<p><img src="olaps.png" width="600px"></p>
<p>The first function, <code><a href="../reference/join_overlap_intersect.html">join_overlap_intersect()</a></code> will return a <code>Ranges</code> object where the start, end, and width coordinates correspond to the amount of any overlap between the left and right input <code>Ranges</code>. It also returns any metadatain the subject range if the subject overlaps the query.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">intersect_rng</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/join_overlap_intersect.html">join_overlap_intersect</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">subject</span><span class="op">)</span>
<span class="va">intersect_rng</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |       key.a       key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##   [1]     chr1       2-4      + |           a           A
##   [2]     chr1       6-7      + |           a           B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="figure">
<img src="an-introduction_files/figure-html/intersect-join-1.png" alt="Intersect Join" width="700"><p class="caption">
Intersect Join
</p>
</div>
<p>The <code><a href="../reference/join_overlap_intersect.html">join_overlap_inner()</a></code> function will return the <code>Ranges</code> in the query that overlap any <code>Ranges</code> in the subject. Like the <code><a href="../reference/join_overlap_intersect.html">join_overlap_intersect()</a></code> function metadata of the subject <code>Range</code> is returned if it overlaps the query.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">inner_rng</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/join_overlap_intersect.html">join_overlap_inner</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">subject</span><span class="op">)</span>
<span class="va">inner_rng</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |       key.a       key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##   [1]     chr1       1-7      + |           a           A
##   [2]     chr1       1-7      + |           a           B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="figure">
<img src="an-introduction_files/figure-html/inner-join-1.png" alt="Inner Join" width="700"><p class="caption">
Inner Join
</p>
</div>
<p>We also provide a convenience method called <code>find_overlaps</code> that computes the same result as <code><a href="../reference/join_overlap_intersect.html">join_overlap_inner()</a></code>.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/find_overlaps.html">find_overlaps</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">subject</span><span class="op">)</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 2 metadata columns:
##       seqnames    ranges strand |       key.a       key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##   [1]     chr1       1-7      + |           a           A
##   [2]     chr1       1-7      + |           a           B
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>The <code><a href="../reference/join_overlap_intersect.html">join_overlap_left()</a></code> method will perform an outer left join.</p>
<p>First any overlaps that are found will be returned similar to <code><a href="../reference/join_overlap_intersect.html">join_overlap_inner()</a></code>. Then any non-overlapping ranges will be returned, with missing values on the metadata columns.</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">left_rng</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/join_overlap_intersect.html">join_overlap_left</a></span><span class="op">(</span><span class="va">query</span>, <span class="va">subject</span><span class="op">)</span>
<span class="va">left_rng</span></code></pre></div>
<pre><code>## GRanges object with 3 ranges and 2 metadata columns:
##       seqnames    ranges strand |       key.a       key.b
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt; &lt;character&gt;
##   [1]     chr1       1-7      + |           a           A
##   [2]     chr1       1-7      + |           a           B
##   [3]     chr1      9-10      - |           b        &lt;NA&gt;
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="figure">
<img src="an-introduction_files/figure-html/olap-left-1.png" alt="Left Join" width="700"><p class="caption">
Left Join
</p>
</div>
<p>Compared with <code><a href="../reference/filter_by_overlaps.html">filter_by_overlaps()</a></code> above, the overlap left join expands the <code>Ranges</code> to give information about each interval on the query <code>Ranges</code> that overlap those on the subject <code>Ranges</code> as well as the intervals on the left that do not overlap any range on the right.</p>
<div id="finding-your-neighbours" class="section level2">
<h2 class="hasAnchor">
<a href="#finding-your-neighbours" class="anchor"></a>Finding your neighbours</h2>
<p>We also provide methods for finding nearest, preceding or following <code>Ranges</code>. Conceptually this is identical to our approach for finding overlaps, except the semantics of the join are different.</p>
<p><img src="neighbours.png" width="600px"></p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/join_nearest.html">join_nearest</a></span><span class="op">(</span><span class="va">ir0</span>, <span class="va">ir1</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width |        gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;numeric&gt;
##   [1]         5         9         5 |  0.780359
##   [2]        10        14         5 |  0.780359
##   [3]        15        19         5 |  0.780359
##   [4]        20        24         5 |  0.780359</code></pre>
<div class="sourceCode" id="cb62"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/join_follow.html">join_follow</a></span><span class="op">(</span><span class="va">ir0</span>, <span class="va">ir1</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width |        gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;numeric&gt;
##   [1]         5         9         5 |  0.777584
##   [2]        10        14         5 |  0.603324
##   [3]        15        19         5 |  0.780359
##   [4]        20        24         5 |  0.780359</code></pre>
<div class="sourceCode" id="cb64"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/join_precede.html">join_precede</a></span><span class="op">(</span><span class="va">ir0</span>, <span class="va">ir1</span><span class="op">)</span> <span class="co"># nothing precedes returns empty `Ranges`</span></code></pre></div>
<pre><code>## IRanges object with 0 ranges and 1 metadata column:
##        start       end     width |        gc
##    &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;numeric&gt;</code></pre>
<div class="sourceCode" id="cb66"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/join_precede.html">join_precede</a></span><span class="op">(</span><span class="va">ir1</span>, <span class="va">ir0</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 5 ranges and 1 metadata column:
##           start       end     width |        gc
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;numeric&gt;
##   [1]         2         4         3 |  0.777584
##   [2]         3         6         4 |  0.827303
##   [3]         4         8         5 |  0.603324
##   [4]         5        10         6 |  0.491232
##   [5]         6        12         7 |  0.780359</code></pre>
</div>
<div id="example-dealing-with-multi-mapping" class="section level2">
<h2 class="hasAnchor">
<a href="#example-dealing-with-multi-mapping" class="anchor"></a>Example: dealing with multi-mapping</h2>
<p>This example is taken from the Bioconductor support <a href="https://support.bioconductor.org/p/100046/">site</a>.</p>
<p>We have two <code>Ranges</code> objects. The first contains single nucleotide positions corresponding to an intensity measurement such as a ChiP-seq experiment, while the other contains coordinates for two genes of interest.</p>
<p>We want to identify which positions in the <code>intensities</code> <code>Ranges</code> overlap the genes, where each row corresponds to a position that overlaps a single gene.</p>
<p>First we create the two <code>Ranges</code> objects</p>
<div class="sourceCode" id="cb68"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">intensities</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"VI"</span>,
                          start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3320</span><span class="op">:</span><span class="fl">3321</span>,<span class="fl">3330</span><span class="op">:</span><span class="fl">3331</span>,<span class="fl">3341</span><span class="op">:</span><span class="fl">3342</span><span class="op">)</span>,
                          width <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/as_iranges.html">as_granges</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">intensities</span></code></pre></div>
<pre><code>## GRanges object with 6 ranges and 0 metadata columns:
##       seqnames    ranges strand
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt;
##   [1]       VI      3320      *
##   [2]       VI      3321      *
##   [3]       VI      3330      *
##   [4]       VI      3331      *
##   [5]       VI      3341      *
##   [6]       VI      3342      *
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<div class="sourceCode" id="cb70"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">genes</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span><span class="op">(</span>seqnames <span class="op">=</span> <span class="st">"VI"</span>,
                    start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3322</span>, <span class="fl">3030</span><span class="op">)</span>,
                    end <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="fl">3846</span>, <span class="fl">3338</span><span class="op">)</span>,
                    gene_id<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="st">"YFL064C"</span>, <span class="st">"YFL065C"</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/as_iranges.html">as_granges</a></span><span class="op">(</span><span class="op">)</span>

<span class="va">genes</span></code></pre></div>
<pre><code>## GRanges object with 2 ranges and 1 metadata column:
##       seqnames    ranges strand |     gene_id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1]       VI 3322-3846      * |     YFL064C
##   [2]       VI 3030-3338      * |     YFL065C
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Now to find where the positions overlap each gene, we can perform an overlap join. This will automatically carry over the gene_id information as well as their coordinates (we can drop those by only selecting the gene_id).</p>
<div class="sourceCode" id="cb72"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">olap</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/join_overlap_intersect.html">join_overlap_inner</a></span><span class="op">(</span><span class="va">intensities</span>, <span class="va">genes</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html">select</a></span><span class="op">(</span><span class="va">gene_id</span><span class="op">)</span>
<span class="va">olap</span></code></pre></div>
<pre><code>## GRanges object with 8 ranges and 1 metadata column:
##       seqnames    ranges strand |     gene_id
##          &lt;Rle&gt; &lt;IRanges&gt;  &lt;Rle&gt; | &lt;character&gt;
##   [1]       VI      3320      * |     YFL065C
##   [2]       VI      3321      * |     YFL065C
##   [3]       VI      3330      * |     YFL065C
##   [4]       VI      3330      * |     YFL064C
##   [5]       VI      3331      * |     YFL065C
##   [6]       VI      3331      * |     YFL064C
##   [7]       VI      3341      * |     YFL064C
##   [8]       VI      3342      * |     YFL064C
##   -------
##   seqinfo: 1 sequence from an unspecified genome; no seqlengths</code></pre>
<p>Several positions match to both genes. We can count them using <code>summarise</code> and grouping by the <code>start</code> position:</p>
<div class="sourceCode" id="cb74"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">olap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/group_by.html">group_by</a></span><span class="op">(</span><span class="va">start</span><span class="op">)</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/summarise.html">summarise</a></span><span class="op">(</span>n <span class="op">=</span> <span class="fu"><a href="../reference/n.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## DataFrame with 6 rows and 2 columns
##       start         n
##   &lt;integer&gt; &lt;integer&gt;
## 1      3320         1
## 2      3321         1
## 3      3330         2
## 4      3331         2
## 5      3341         1
## 6      3342         1</code></pre>
</div>
<div id="grouping-by-overlaps" class="section level2">
<h2 class="hasAnchor">
<a href="#grouping-by-overlaps" class="anchor"></a>Grouping by overlaps</h2>
<p>Itâ€™s also possible to group by overlaps. Using this approach we can count the number of overlaps that are greater than 0.</p>
<div class="sourceCode" id="cb76"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grp_by_olap</span> <span class="op">&lt;-</span> <span class="va">ir0</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="../reference/find_overlaps.html">group_by_overlaps</a></span><span class="op">(</span><span class="va">ir1</span><span class="op">)</span>
<span class="va">grp_by_olap</span></code></pre></div>
<pre><code>## IRanges object with 6 ranges and 2 metadata columns:
## Groups: query [2]
##           start       end     width |        gc     query
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;numeric&gt; &lt;integer&gt;
##   [1]         5         9         5 |  0.827303         1
##   [2]         5         9         5 |  0.603324         1
##   [3]         5         9         5 |  0.491232         1
##   [4]         5         9         5 |  0.780359         1
##   [5]        10        14         5 |  0.491232         2
##   [6]        10        14         5 |  0.780359         2</code></pre>
<div class="sourceCode" id="cb78"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">grp_by_olap</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fu"><a href="../reference/n.html">n</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 6 ranges and 3 metadata columns:
## Groups: query [2]
##           start       end     width |        gc     query n_overlaps
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; | &lt;numeric&gt; &lt;integer&gt;  &lt;integer&gt;
##   [1]         5         9         5 |  0.827303         1          4
##   [2]         5         9         5 |  0.603324         1          4
##   [3]         5         9         5 |  0.491232         1          4
##   [4]         5         9         5 |  0.780359         1          4
##   [5]        10        14         5 |  0.491232         2          2
##   [6]        10        14         5 |  0.780359         2          2</code></pre>
<p>Of course we can also add overlap counts via the <code><a href="../reference/count_overlaps.html">count_overlaps()</a></code> function.</p>
<div class="sourceCode" id="cb80"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">ir0</span> <span class="op">%&gt;%</span>
  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html">mutate</a></span><span class="op">(</span>n_overlaps <span class="op">=</span> <span class="fu"><a href="../reference/count_overlaps.html">count_overlaps</a></span><span class="op">(</span><span class="va">.</span>, <span class="va">ir1</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## IRanges object with 4 ranges and 1 metadata column:
##           start       end     width | n_overlaps
##       &lt;integer&gt; &lt;integer&gt; &lt;integer&gt; |  &lt;integer&gt;
##   [1]         5         9         5 |          4
##   [2]        10        14         5 |          2
##   [3]        15        19         5 |          0
##   [4]        20        24         5 |          0</code></pre>
</div>
</div>
<div id="data-importoutput" class="section level1">
<h1 class="hasAnchor">
<a href="#data-importoutput" class="anchor"></a>Data Import/Output</h1>
<p>We provide convenience functions via <code>rtracklayer</code> and <code>GenomicAlignments</code> for reading/writing the following data formats to/from <code>Ranges</code> objects.</p>
<table class="table">
<thead><tr class="header">
<th>
<code>plyranges</code> functions</th>
<th>File Format</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code><a href="../reference/read_bam.html">read_bam()</a></code></td>
<td>BAM</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/read_bed.html">read_bed()</a></code>/<code><a href="../reference/write_bed.html">write_bed()</a></code>
</td>
<td>BED</td>
</tr>
<tr class="odd">
<td>
<code><a href="../reference/read_bed.html">read_bed_graph()</a></code>/ <code><a href="../reference/write_bed.html">write_bed_graph()</a></code>
</td>
<td>BEDGraph</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/read_bed.html">read_narrowpeaks()</a></code>/<code><a href="../reference/write_bed.html">write_narrowpeaks()</a></code>
</td>
<td>narrowPeaks</td>
</tr>
<tr class="odd">
<td>
<code><a href="../reference/read_gff.html">read_gff()</a></code> / <code><a href="../reference/write_gff.html">write_gff()</a></code>
</td>
<td>GFF(1-3)/ GTF</td>
</tr>
<tr class="even">
<td>
<code><a href="../reference/read_bigwig.html">read_bigwig()</a></code> / <code><a href="../reference/write_bigwig.html">write_bigwig()</a></code>
</td>
<td>BigWig</td>
</tr>
<tr class="odd">
<td>
<code><a href="../reference/read_wig.html">read_wig()</a></code> /<code><a href="../reference/write_wig.html">write_wig()</a></code>
</td>
<td>Wig</td>
</tr>
</tbody>
</table>
</div>
<div id="learning-more" class="section level1">
<h1 class="hasAnchor">
<a href="#learning-more" class="anchor"></a>Learning more</h1>
<p>There are many other resources and workshops available to learn to use <code>plyranges</code> and related Bioconductor packages, especially for more realistic analyses than the ones covered here:</p>
<ul>
<li>The <a href="https://bioconductor.org/packages/release/workflows/html/fluentGenomics.html">fluentGenomics</a> workflow package is an end-to-end workflow package for integrating differential expression results with differential accessibility results.</li>
<li>The <a href="https://bioconductor.github.io/BiocWorkshops/fluent-genomic-data-analysis-with-plyranges.html">Bioc 2018 Workshop book</a> has worked examples of using <code>plyranges</code> to analyse publicly available genomics data.</li>
<li>The <a href="https://github.com/sa-lee/plyrangesWorkshops">extended vignette in the plyrangesWorkshops package</a> has a detailed walk through of using plyranges for coverage analysis.</li>
<li>The <a href="https://github.com/mikelove/plyrangesTximetaCaseStudy">case study</a> by Michael Love using plyranges with <a href="https://bioconductor.org/packages/release/bioc/html/tximeta.html">tximeta</a> to follow up on interesting hits from a combined RNA-seq and ATAC-seq analysis.</li>
<li>The <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-018-1597-8">journal article</a> (<a href="https://www.biorxiv.org/content/early/2018/05/23/327841">preprint here</a>) has details about the overall philosophy and design of plyranges.</li>
</ul>
</div>
<div id="appendix" class="section level1">
<h1 class="hasAnchor">
<a href="#appendix" class="anchor"></a>Appendix</h1>
<div class="sourceCode" id="cb82"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R Under development (unstable) (2021-04-15 r80175)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.2 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] stats4    parallel  stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
## [1] ggplot2_3.3.3        plyranges_1.11.2     GenomicRanges_1.43.4
## [4] GenomeInfoDb_1.27.11 IRanges_2.25.10      S4Vectors_0.29.18   
## [7] BiocGenerics_0.37.5  BiocStyle_2.19.2    
## 
## loaded via a namespace (and not attached):
##  [1] MatrixGenerics_1.3.1        Biobase_2.51.0             
##  [3] assertthat_0.2.1            BiocManager_1.30.12        
##  [5] highr_0.9                   GenomeInfoDbData_1.2.4     
##  [7] Rsamtools_2.7.2             yaml_2.2.1                 
##  [9] pillar_1.6.0                lattice_0.20-41            
## [11] glue_1.4.2                  digest_0.6.27              
## [13] XVector_0.31.1              colorspace_2.0-1           
## [15] htmltools_0.5.1.1           Matrix_1.3-2               
## [17] XML_3.99-0.6                pkgconfig_2.0.3            
## [19] bookdown_0.22               zlibbioc_1.37.0            
## [21] purrr_0.3.4                 scales_1.1.1               
## [23] BiocParallel_1.25.5         tibble_3.1.1               
## [25] farver_2.1.0                generics_0.1.0             
## [27] ellipsis_0.3.2              withr_2.4.2                
## [29] cachem_1.0.4                SummarizedExperiment_1.21.3
## [31] magrittr_2.0.1              crayon_1.4.1               
## [33] memoise_2.0.0               evaluate_0.14              
## [35] fs_1.5.0                    fansi_0.4.2                
## [37] textshaping_0.3.3           tools_4.1.0                
## [39] BiocIO_1.1.2                lifecycle_1.0.0            
## [41] matrixStats_0.58.0          stringr_1.4.0              
## [43] munsell_0.5.0               DelayedArray_0.17.13       
## [45] Biostrings_2.59.2           compiler_4.1.0             
## [47] pkgdown_1.6.1               systemfonts_1.0.1          
## [49] rlang_0.4.11                grid_4.1.0                 
## [51] RCurl_1.98-1.3              rjson_0.2.20               
## [53] labeling_0.4.2              bitops_1.0-7               
## [55] rmarkdown_2.7               restfulr_0.0.13            
## [57] gtable_0.3.0                DBI_1.1.1                  
## [59] R6_2.5.0                    GenomicAlignments_1.27.2   
## [61] knitr_1.33                  dplyr_1.0.6                
## [63] rtracklayer_1.51.5          fastmap_1.1.0              
## [65] utf8_1.2.1                  rprojroot_2.0.2            
## [67] ragg_1.1.2                  desc_1.3.0                 
## [69] stringi_1.5.3               vctrs_0.3.8                
## [71] tidyselect_1.1.1            xfun_0.22</code></pre>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://stuartlee.org">Stuart Lee</a>, <a href="https://github.com/lawremi">Michael Lawrence</a>, Dianne Cook.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
